plugins/formal-verification/formver.core/src/org/jetbrains/kotlin/formver/Purity/ExprPurityVisitor.kt

```diff
@@ -0,0 +1,106 @@
+import org.jetbrains.kotlin.formver.embeddings.expression.Declare
+import org.jetbrains.kotlin.formver.embeddings.expression.Elvis
+import org.jetbrains.kotlin.formver.embeddings.expression.EqCmp
+import org.jetbrains.kotlin.formver.embeddings.expression.ErrorExp
+import org.jetbrains.kotlin.formver.embeddings.expression.ExpEmbedding
+import org.jetbrains.kotlin.formver.embeddings.expression.FieldAccess
+import org.jetbrains.kotlin.formver.embeddings.expression.FieldAccessPermissions
+import org.jetbrains.kotlin.formver.embeddings.expression.FieldModification
+import org.jetbrains.kotlin.formver.embeddings.expression.ForAllEmbedding
+import org.jetbrains.kotlin.formver.embeddings.expression.FunctionExp
+import org.jetbrains.kotlin.formver.embeddings.expression.Goto
+import org.jetbrains.kotlin.formver.embeddings.expression.GotoChainNode
+import org.jetbrains.kotlin.formver.embeddings.expression.If
+import org.jetbrains.kotlin.formver.embeddings.expression.InhaleDirect
+import org.jetbrains.kotlin.formver.embeddings.expression.InhaleInvariants
+import org.jetbrains.kotlin.formver.embeddings.expression.InjectionBasedExpEmbedding
+import org.jetbrains.kotlin.formver.embeddings.expression.InvokeFunctionObject
+import org.jetbrains.kotlin.formver.embeddings.expression.Is
+import org.jetbrains.kotlin.formver.embeddings.expression.LabelExp
+import org.jetbrains.kotlin.formver.embeddings.expression.LambdaExp
+import org.jetbrains.kotlin.formver.embeddings.expression.MethodCall
+import org.jetbrains.kotlin.formver.embeddings.expression.NeCmp
+import org.jetbrains.kotlin.formver.embeddings.expression.NonDeterministically
+import org.jetbrains.kotlin.formver.embeddings.expression.Old
+import org.jetbrains.kotlin.formver.embeddings.expression.PredicateAccessPermissions
+import org.jetbrains.kotlin.formver.embeddings.expression.PrimitiveFieldAccess
+import org.jetbrains.kotlin.formver.embeddings.expression.PureExpEmbedding
+import org.jetbrains.kotlin.formver.embeddings.expression.SafeCast
+import org.jetbrains.kotlin.formver.embeddings.expression.SequentialAnd
+import org.jetbrains.kotlin.formver.embeddings.expression.SequentialOr
+import org.jetbrains.kotlin.formver.embeddings.expression.Shared
+import org.jetbrains.kotlin.formver.embeddings.expression.SharingContext
+import org.jetbrains.kotlin.formver.embeddings.expression.UnaryOperatorExpEmbedding
+import org.jetbrains.kotlin.formver.embeddings.expression.UnitLit
+import org.jetbrains.kotlin.formver.embeddings.expression.While
+import org.jetbrains.kotlin.formver.embeddings.expression.WithPosition
+
+internal object ExprPurityVisitor : ExpVisitor<Boolean> {
+
+    /* ————— pure nodes ————— */
+    override fun visitPureExpEmbedding(e: PureExpEmbedding) = true
+    override fun visitUnitLit(e: UnitLit) = true
+
+    /* ————— structural nodes without side effects ————— */
+    override fun visitBlock(e: Block) = e.recurse(this)
+    override fun visitIf(e: If) = e.recurse(this)
```

Comment:
I would mark this as non-pure at this point since we cannot generate pure code for it. (idem for some others)